File: assignments/assignment-2-solution/src-csharp/TeaShop/UserInterface/Application.cs
using Assignment2Solution.Domain.Inventory;
using Assignment2Solution.Domain.Query;
using Assignment2Solution.UserInterface.PaymentMethodGenerator;
using Assignment2Solution.UserInterface.Query;

namespace Assignment2Solution.UserInterface;

/// <summary>
///     The main application class that coordinates the user interface and domain logic.
/// </summary>
public sealed class Application
{
    private readonly TextReader _input;
    private readonly TextWriter _output;
    private readonly IReadOnlyList<IPaymentStrategyGenerator> _paymentMethods;

    private readonly QueryBuilder _queryBuilder;
    private readonly QueryOutputWriter _queryOutputWriter;
    private readonly InventoryRepository _repository;

    /// <summary>
    ///     Initializes a new instance of the <see cref="Application" /> class.
    /// </summary>
    /// <param name="input">The text reader for user input.</param>
    /// <param name="output">The text writer for application output.</param>
    public Application(TextReader input, TextWriter output)
    {
        _input = input ?? throw new ArgumentNullException(nameof(input));
        _output = output ?? throw new ArgumentNullException(nameof(output));

        _repository = new InventoryRepository();
        _queryBuilder = new QueryBuilder(_repository, _input, _output);
        _queryOutputWriter = new QueryOutputWriter(_output);

        _paymentMethods = PaymentStrategyGeneratorListFactory.Get();
    }

    /// <summary>
    ///     Runs the application loop.
    /// </summary>
    public void Run()
    {
        DisplayWelcomeMessage();

        while (true)
        {
            var query = _queryBuilder.Build();
            var output = QueryOutput.From(query);
            _queryOutputWriter.Write(output);
            _output.WriteLine();

            if (output.Items.Count == 0)
                _output.WriteLine("*** Nothing matched your search ***");
            else
                ProcessPurchase(output);

            _output.WriteLine();
            var searchForMore = ReadYesNo("Search for more tea? (Y/N, default Y): ", true);
            if (!searchForMore) break;
            _output.WriteLine();
        }
    }

    private void DisplayWelcomeMessage()
    {
        _output.WriteLine("WELCOME TO JEFF'S TEA SHOP");
        _output.WriteLine();
        _output.WriteLine("Complete the prompts to search our selection of fine teas.");
        _output.WriteLine();
    }

    private void ProcessPurchase(QueryOutput output)
    {
        ArgumentNullException.ThrowIfNull(output);

        if (output.Items.Count == 0)
        {
            _output.WriteLine("!!! No items available to purchase.");
            return;
        }

        while (true)
        {
            var itemSelectionPrompt = $"1-{output.Items.Count} or 0 to continue (default)";
            _output.Write($"Purchase an item? Enter item number {itemSelectionPrompt}: ");
            var input = _input.ReadLine();

            if (string.IsNullOrWhiteSpace(input)) return;

            if (!int.TryParse(input.Trim(), out var index) || index < 0 || index > output.Items.Count)
            {
                _output.WriteLine("!!! Invalid item number. Try again.");
                continue;
            }

            if (index == 0) return;

            var selected = output.Items[index - 1];

            if (selected.Quantity <= 0)
            {
                _output.WriteLine($"!!! Sorry, there is inventory available for {selected.Name}.");
                return;
            }

            _output.Write($"Quantity for \"{selected.Name}\" (1-{selected.Quantity}): ");
            var qtyInput = _input.ReadLine();
            if (!int.TryParse(qtyInput?.Trim(), out var quantity) || quantity < 1 || quantity > selected.Quantity)
            {
                _output.WriteLine("!!! Invalid quantity.");
                continue;
            }

            var totalPrice = selected.Price * quantity;
            _output.WriteLine($"*** Total Price: {totalPrice:C}");

            ProcessCheckout(selected, quantity);
            break;
        }
    }

    private void ProcessCheckout(QueriedInventoryItem item, int quantity)
    {
        _output.WriteLine("*** Choose a payment method:");
        for (var i = 0; i < _paymentMethods.Count; i++) _output.WriteLine($"{i + 1}. {_paymentMethods[i].Name}");

        _output.Write("Selection: ");

        var choice = _input.ReadLine();
        if (!int.TryParse(choice, out var index) || index < 1 || index > _paymentMethods.Count)
        {
            _output.WriteLine("!!! Invalid payment method selection. Checkout cancelled.");
            return;
        }

        var strategy = _paymentMethods[index - 1].CreateStrategy(_input, _output);
        strategy.Checkout(item, quantity, _output);

        // Decrease inventory quantity (using negative value for decrease)
        _repository.UpdateQuantity(item.InventoryItemId, -1 * quantity);

        var purchaseDesc = $"{quantity} packages of {item.Name}";
        _output.WriteLine($"*** Purchase complete. Your {purchaseDesc} is on the way ***");
    }

    private bool ReadYesNo(string prompt, bool defaultValue)
    {
        while (true)
        {
            _output.Write(prompt);
            var input = _input.ReadLine();

            if (string.IsNullOrWhiteSpace(input)) return defaultValue;

            if (input.Trim().Equals("Y", StringComparison.OrdinalIgnoreCase))
                return true;
            if (input.Trim().Equals("N", StringComparison.OrdinalIgnoreCase))
                return false;

            _output.WriteLine("Please enter Y or N.");
        }
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/userinterface/Application.java
package assignment2solution.userinterface;

import assignment2solution.domain.inventory.InventoryRepository;
import assignment2solution.domain.inventoryquery.QueriedInventoryItem;
import assignment2solution.userinterface.paymentbuilder.IPaymentBuilder;
import assignment2solution.userinterface.paymentbuilder.PaymentBuilderListFactory;
import assignment2solution.userinterface.querybuilder.InventoryQueryBuilder;
import assignment2solution.userinterface.querybuilder.InventoryQueryOutput;
import assignment2solution.userinterface.querybuilder.InventoryQueryOutputWriter;

import java.io.BufferedReader;
import java.io.Reader;
import java.io.Writer;
import java.io.PrintWriter;
import java.math.BigDecimal;
import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

/**
 * The main application class that coordinates the user interface and domain logic.
 */
public final class Application {
    private final BufferedReader input;
    private final PrintWriter output;
    private final InventoryRepository repository;
    private final InventoryQueryBuilder inventoryQueryBuilder;
    private final InventoryQueryOutputWriter inventoryQueryOutputWriter;
    private final List<IPaymentBuilder> paymentMethods;
    private final NumberFormat currencyFormatter = NumberFormat.getCurrencyInstance(Locale.US);

    /**
     * Initializes a new instance of the {@link Application} class.
     *
     * @param input  The text reader for user input.
     * @param output The text writer for application output.
     */
    public Application(Reader input, Writer output) {
        Objects.requireNonNull(input, "input");
        Objects.requireNonNull(output, "output");

        this.input = new BufferedReader(input);
        this.output = new PrintWriter(output, true);
        this.repository = new InventoryRepository();
        this.inventoryQueryBuilder = new InventoryQueryBuilder(repository, this.input, this.output);
        this.inventoryQueryOutputWriter = new InventoryQueryOutputWriter(this.output);
        this.paymentMethods = PaymentBuilderListFactory.get();
    }

    /**
     * Runs the application loop.
     */
    public void run() {
        displayWelcomeMessage();

        while (true) {
            var query = inventoryQueryBuilder.build();
            var output = InventoryQueryOutput.from(query);
            inventoryQueryOutputWriter.write(output);
            this.output.println();

            if (output.items().isEmpty()) {
                this.output.println("*** Nothing matched your search ***");
            } else {
                processPurchase(output);
            }

            this.output.println();
            var searchForMore = readYesNo("Search for more tea? (Y/N, default Y): ", true);
            if (!searchForMore) {
                break;
            }
            this.output.println();
        }
    }

    private void displayWelcomeMessage() {
        output.println("WELCOME TO JEFF'S TEA SHOP");
        output.println();
        output.println("Complete the prompts to search our selection of fine teas.");
        output.println();
    }

    private void processPurchase(InventoryQueryOutput output) {
        Objects.requireNonNull(output, "output");

        if (output.items().isEmpty()) {
            this.output.println("!!! No items available to purchase.");
            return;
        }

        while (true) {
            var itemSelectionPrompt = "1-" + output.items().size() + " or 0 to continue (default)";
            printPrompt("Purchase an item? Enter item number " + itemSelectionPrompt + ": ");
            var inputLine = readLine();

            if (inputLine == null || inputLine.trim().isEmpty()) {
                return;
            }

            int index;
            try {
                index = Integer.parseInt(inputLine.trim());
            } catch (NumberFormatException ex) {
                this.output.println("!!! Invalid item number. Try again.");
                continue;
            }

            if (index < 0 || index > output.items().size()) {
                this.output.println("!!! Invalid item number. Try again.");
                continue;
            }

            if (index == 0) {
                return;
            }

            var selected = output.items().get(index - 1);

            if (selected.getQuantity() <= 0) {
                this.output.println("!!! Sorry, there is inventory available for " + selected.getName() + ".");
                return;
            }

            printPrompt("Quantity for \"" + selected.getName() + "\" (1-" + selected.getQuantity() + "): ");
            var qtyInput = readLine();
            int quantity;
            try {
                quantity = Integer.parseInt(qtyInput == null ? "" : qtyInput.trim());
            } catch (NumberFormatException ex) {
                this.output.println("!!! Invalid quantity.");
                continue;
            }
            if (quantity < 1 || quantity > selected.getQuantity()) {
                this.output.println("!!! Invalid quantity.");
                continue;
            }

            var totalPrice = selected.getPrice().multiply(BigDecimal.valueOf(quantity));
            this.output.println("*** Total Price: " + currencyFormatter.format(totalPrice));

            processCheckout(selected, quantity);
            break;
        }
    }

    private void processCheckout(QueriedInventoryItem item, int quantity) {
        this.output.println("*** Choose a payment method:");
        for (int i = 0; i < paymentMethods.size(); i++) {
            this.output.println((i + 1) + ". " + paymentMethods.get(i).getName());
        }

        printPrompt("Selection: ");

        var choice = readLine();
        int index;
        try {
            index = Integer.parseInt(choice == null ? "" : choice.trim());
        } catch (NumberFormatException ex) {
            this.output.println("!!! Invalid payment method selection. Checkout cancelled.");
            return;
        }

        if (index < 1 || index > paymentMethods.size()) {
            this.output.println("!!! Invalid payment method selection. Checkout cancelled.");
            return;
        }

        var strategy = paymentMethods.get(index - 1).createStrategy(input, output);
        strategy.checkout(item, quantity, output);

        // Decrease inventory quantity (using negative value for decrease)
        repository.updateQuantity(item.getInventoryItemId(), -1 * quantity);

        var purchaseDesc = quantity + " packages of " + item.getName();
        this.output.println("*** Purchase complete. Your " + purchaseDesc + " is on the way ***");
    }

    private boolean readYesNo(String prompt, boolean defaultValue) {
        while (true) {
            printPrompt(prompt);
            var inputLine = readLine();

            if (inputLine == null || inputLine.trim().isEmpty()) {
                return defaultValue;
            }

            var trimmed = inputLine.trim();
            if (trimmed.equalsIgnoreCase("Y")) {
                return true;
            }
            if (trimmed.equalsIgnoreCase("N")) {
                return false;
            }

            this.output.println("Please enter Y or N.");
        }
    }

    private String readLine() {
        try {
            return input.readLine();
        } catch (Exception ex) {
            return null;
        }
    }

    private void printPrompt(String prompt) {
        this.output.print(prompt);
        this.output.flush();
    }
}
